File: [K]graph/[K]graph-bfs/994-rotting-oranges.py
from collections import deque
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        rows, cols = len(grid), len(grid[0])
        visited = set()
        queue = deque([])
        fresh = 0
        for r in range(rows):
            for c in range(cols):
                if grid[r][c] == 1:
                    fresh += 1
                elif grid[r][c] == 2:
                    queue.append((r, c, 0))
                    visited.add((r, c))
        if fresh == 0:
            return 0

        while queue:
            r, c, minute = queue.popleft()
            if grid[r][c] == 1:
                fresh -= 1
                if fresh == 0:
                    return minute
            for next_r, next_c in [(r+1, c), (r-1, c), (r, c+1), (r, c-1)]:
                if next_r not in range(rows) or next_c not in range(cols) or (next_r, next_c) in visited or grid[next_r][next_c] != 1:
                    continue
                queue.append((next_r, next_c, minute + 1))
                visited.add((next_r, next_c))
            
        return - 1
    
# time O(RC)
# space O(RC)
# using graph and bfs with multiple sources

File: [K]graph/[K]graph-bfs/994-rotting-oranges.java
class Solution {
    public int orangesRotting(int[][] grid) {
        int rows = grid.length;
        int cols = grid[0].length;
        int fresh = 0;
        int rotten = 0;
        ArrayDeque<int[]> queue = new ArrayDeque<>();
        HashSet<Integer> visited = new HashSet<>();
        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) {
                if (grid[r][c] == 1) {
                    fresh += 1;
                }
                else if (grid[r][c] == 2) {
                    rotten += 1;
                    queue.offer(new int[]{r, c, 0});
                    visited.add(r * cols + c);
                }
            }
        }
        if (fresh == 0) {
            return 0;
        }
        if (rotten == 0) {
            return - 1;
        }
        int res = 0;
        while (!queue.isEmpty()) {
            int[] rc = queue.poll();
            int r = rc[0];
            int c = rc[1];
            int step = rc[2];
            res = step;
            if (grid[r][c] == 1) {
                fresh -= 1;
            }
            for (int[] nextrc: new int[][]{{r+1, c}, {r-1, c}, {r, c+1}, {r, c-1}}) {
                int next_r = nextrc[0];
                int next_c = nextrc[1];
                if (0 <= next_r && next_r < rows && 0 <= next_c && next_c < cols && !visited.contains(next_r * cols + next_c) && grid[next_r][next_c] == 1) {
                    queue.offer(new int[]{next_r, next_c, step + 1});
                    visited.add(next_r * cols + next_c);
                }
            }
        }
        return fresh == 0 ? res : - 1;
    }
}

// time O(RC)
// space O(RC)
// using graph and bfs with multiple sources